### ISSUE TITLE
heif: Unsupported feature: Unsupported codec

### ISSUE THREAD

**Issue by enoixer** [nsg/immich-distribution#137]
### Channel

candidate

### Linux Distribution

Other Linux (please specify)

### What happened?

**The OS that Immich Server is running on:**

Debian 12 Proxmox Container

**Version of Immich Server**

1.88.2

**Reproduction steps:**

Upload one iOS Live photo (HEIF format).

**The bug:**

In iOS HEIF image uploads, the immich-microservices service cannot generate a thumbnail logging this error:

heif: Unsupported feature: Unsupported codec (4.3000)#033[39m

This bug normally occurrs when the libvips library used by sharp library is not compiled with HEIF support.

In the old immich releases like 10.85.0 for example, libvips was compiled using the build-libvips.sh script in order to solve this bug, but I can see this script is not used anymore in the last release.

The are one immich issue related with this bug:

https://github.com/immich-app/immich/issues/3943

Maybe, you can check if the libvips used in your snap is compiled with heif support.

Thank you very much for create this immich-distribution!

### Provide logs from journalctl

```bash
2023-11-29T19:05:34.777402+01:00 photos immich-distribution.immich-microservices[1170]: #033[31m[Nest] 1170  - #033[39m11/29/2023, 7:05:34 PM #033[31m  ERROR#033[39m #033[38;5;3m[JobService] #033[39m#033[31mObject:#033[39m
2023-11-29T19:05:34.777422+01:00 photos immich-distribution.immich-microservices[1170]: {
2023-11-29T19:05:34.777443+01:00 photos immich-distribution.immich-microservices[1170]:   "id": "3732b314-b5bd-4766-877c-d13563d78a53",
2023-11-29T19:05:34.777464+01:00 photos immich-distribution.immich-microservices[1170]:   "source": "upload"
2023-11-29T19:05:34.777483+01:00 photos immich-distribution.immich-microservices[1170]: }
2023-11-29T19:05:34.970418+01:00 photos immich-distribution.immich-microservices[1170]: #033[31m[Nest] 1170  - #033[39m11/29/2023, 7:05:34 PM #033[31m  ERROR#033[39m #033[38;5;3m[JobService] #033[39m#033[31mUnable to run job handler (thumbnailGeneration/generate-jpeg-thumbnail): Error: upload/library/admin/2023/2023-11-23/IMG_0788.HEIC: bad seek to 560106
2023-11-29T19:05:34.970533+01:00 photos immich-distribution.immich-microservices[1170]: heif: Unsupported feature: Unsupported codec (4.3000)#033[39m
2023-11-29T19:05:34.970580+01:00 photos immich-distribution.immich-microservices[1170]: #033[31m[Nest] 1170  - #033[39m11/29/2023, 7:05:34 PM #033[31m  ERROR#033[39m #033[38;5;3m[JobService] #033[39m#033[31mError: upload/library/admin/2023/2023-11-23/IMG_0788.HEIC: bad seek to 560106
2023-11-29T19:05:34.970632+01:00 photos immich-distribution.immich-microservices[1170]: heif: Unsupported feature: Unsupported codec (4.3000)#033[39m
2023-11-29T19:05:34.970674+01:00 photos immich-distribution.immich-microservices[1170]: #033[31m[Nest] 1170  - #033[39m11/29/2023, 7:05:34 PM #033[31m  ERROR#033[39m #033[38;5;3m[JobService] #033[39m#033[31mObject:#033[39m
2023-11-29T19:05:34.970698+01:00 photos immich-distribution.immich-microservices[1170]: {
2023-11-29T19:05:34.970720+01:00 photos immich-distribution.immich-microservices[1170]:   "id": "933a75bf-e4a6-4687-8db6-2082ec7c8dca",
2023-11-29T19:05:34.970752+01:00 photos immich-distribution.immich-microservices[1170]:   "source": "upload"
2023-11-29T19:05:34.970775+01:00 photos immich-distribution.immich-microservices[1170]: }
```

**Comment by nsg** [comment]
I build it against `libheif-dev` and I bundle `libheif1` in the snap. I _think_ the library version shipped in core20 is to old https://github.com/libvips/libvips/issues/1902. I will investigate it closer.

**Comment by enoixer** [comment]
Ok, thank you very mucho @nsg !

**Comment by nsg** [comment]
This was a tricky one

* https://github.com/nsg/immich-distribution/pull/140/commits/e84048baac0d4a68fec786416184f4e8466bfa9b builds libheif, libde265 and x265 from source. Libvips pick that one up and links against it (verified with ldd).
* Updated LD_LIBRARY_PATH https://github.com/nsg/immich-distribution/pull/140/commits/bb2f8caa5269899b6c8643fcf8b31569d8e60f51

At this point I was able to use `vips` and `heif-convert` to convert files. But sharp refused to pick up libheif support. I finally found this thread https://github.com/lovell/sharp/issues/3680 that suggested that the new plugin based architecture that newer versions of libheif uses by default causes problems for sharp so I disabled it.

* Disable plugin based system for libheif in https://github.com/nsg/immich-distribution/pull/140/commits/5cdd20e36894916af8aa5ac295e93c52ae09ccff

When I was going to test it I was hit with [this bug](https://github.com/orgs/nodejs/discussions/49784), I solved it with a bump to Node v20.10, it's a LTS release so it's probably a good thing to use anyway.

* Bump Node version in https://github.com/nsg/immich-distribution/pull/140/commits/6716d9df8a1a4fb34a0441bd0e2f3251024dee0

Not sure if everyting works or not and it's to late to test everything, tomorrow maybe!

**Comment by enoixer** [comment]
Hi @nsg,

I was looking in the docker immich 1.85.0 and I can see the libvips uses vips-modules I understand this modules are the plugins as you said?:

```
ls -l /usr/local/lib/vips-modules-8.14
```

```
-rwxr-xr-x 1 root root 57032 Nov 28 23:11 vips-heif.so
-rwxr-xr-x 1 root root 45608 Nov 28 23:11 vips-jxl.so
-rwxr-xr-x 1 root root 56384 Nov 28 23:11 vips-magick.so
```

Maybe the sharp version used in the docker immich is prepared for uses the libvips library with modules and/or plugins?.

@ngs, thank you very much for you hard work in order to solve this issue!

**Comment by nsg** [comment]
Sorry for several weeks of silence,

I have been tinkering with this behind the scenes, trying to find time between some travel, a few events, Christmas, visiting family, work, and I was also sick for a week. So to summarize, a busy month! :grin: 

I'm not completely done yet, but I have more or less ripped everything apart to figure this one out. I consider HEIC images really important so felt that I needed to understand, and solve this one properly.

I have now a snap build that runs this code:
```js
const sharp = require('/snap/immich-distribution/current/lib/node_modules/immich/node_modules/sharp');

async function resize() {
    await sharp("$SNAP_USER_DATA/zfs-broken.png").toFile("$SNAP_USER_DATA/zfs-broken.jpg");
    await sharp("$SNAP_USER_DATA/dwsample.heic").heif({ compression: 'hevc' }).toFile("$SNAP_USER_DATA/dwsample.png");
    await sharp("$SNAP_USER_DATA/dwsample.heic").toFile("$SNAP_USER_DATA/dwsample.2.png");
}

resize();
```

I'm now in the progress to adding Immich back to the snap again, I have lifted and improved a lot of things, like for example, I have updated from `core20` (based on Ubuntu 20.04) to `core22` (22.04) and a lot of refactors and general improvements. So... it's possible that there is regressions and Immich build issues now, but that's probably a problem for next year.

Have a Happy New Year, and one again, thank you for your issue and helpful messages. I really hope I will be able to get this one out in a few days and then all normal Immich updates after that.

**Comment by nsg** [comment]
The test file that you provided https://www.dwsamplefiles.com/?dl_id=25 (`dwsample-heic-640.heic`) renders in nicely Immich Web.

```
$ file dwsample-heic-640.heic
dwsample-heic-640.heic: ISO Media, HEIF Image HEVC Main or Main Still Picture Profile
```

![image](https://github.com/nsg/immich-distribution/assets/16388/11cea1ce-644f-4821-89d4-293dc136223a)

I have also tried a few other sample images, one taken with an iPhone 12 Pro.

**Comment by nsg** [comment]
The fixes are part of the [v1.89](https://github.com/nsg/immich-distribution/issues/138) release.
